File: DijkstrasAlgorithm/DijkstrasAlgorithm/Vertex.cs
namespace DijkstrasAlgorithm
{
    using System.Linq;

    public class Vertex
    {
        private readonly string _Name;

        public Vertex(string name)
        {
            _Name = name;
            Cost = int.MaxValue;
            Parent = null;
        }

        public Edge AdjacentEdge { get; private set; }

        public int Cost { get; set; }

        public string Name
        {
            get
            {
                return _Name;
            }
        }

        public Vertex Parent { get; set; }

        public bool IsAdjacentTo(Vertex v, Edge[] edges)
        {
            var connectingEdges = from e in edges
                                  where e.ConnectedTo(v)
                                  && e.ConnectedTo(this)
                                  orderby e.Cost
                                  select e;
            this.AdjacentEdge = connectingEdges.FirstOrDefault();
            return this.AdjacentEdge != null;
        }

        public override string ToString()
        {
            string p = Parent != null ? Parent.Name : "-";
            string c = Cost < int.MaxValue ? Cost.ToString() : "-";
            return string.Format("{0} ({1}, {2})", Name, p, c);
        }
    }
}

File: DijkstrasAlgorithm/dijkstra/vertex.cpp
#include "vertex.h"
#include "edge.h"

using namespace std;

Vertex::Vertex( std::string name ) :
    Cost( -1 ),
    Id( name ),
    AdjacentEdge( NULL ),
    Parent( NULL )
{
}

Vertex::Vertex( const Vertex &v ) :
    Cost( v.Cost ),
    Id( v.Id ),
    AdjacentEdge( NULL ),
    Parent( NULL )
{
    if ( v.AdjacentEdge != NULL )
    {
        AdjacentEdge = new Edge( *v.AdjacentEdge );
    }

    Parent = v.Parent;
}

Vertex::~Vertex()
{
    delete AdjacentEdge;
}

Vertex *Vertex::GetParent() const
{
    return Parent;
}

bool Vertex::IsAdjacentTo( const Vertex &v, std::vector<Edge>& edges )
{
    AdjacentEdge = NULL;

    for ( std::vector<Edge>::iterator e = edges.begin();
            e != edges.end();
            ++e )
    {
        if ( e->ConnectedTo( v ) && e->ConnectedTo( *this ) )
        {
            if ( AdjacentEdge == NULL || e->Cost < AdjacentEdge->Cost )
            {
                if ( AdjacentEdge != NULL )
                {
                    delete AdjacentEdge;
                }

                AdjacentEdge = new Edge( *e );
            }
        }
    }

    AdjacentEdge != NULL;
}

std::string Vertex::ToString()
{
    Vertex *parent = GetParent();
    string p = parent != NULL ? parent->Id : "-";
    string c;

    if ( -1 < Cost )
    {
        stringstream t;
        t << Cost;
        c = t.str();
    }
    else
    {
        c = "-";
    }

    return Id + "(" + p + ", " + c + ")";
}

Vertex &Vertex::operator=( const Vertex &v )
{
    if ( this == &v )
    {
        return *this;
    }

    int cost = v.Cost;
    string name = v.Id;
    Edge *e = NULL;
    Vertex *p = v.Parent;

    if ( v.AdjacentEdge != NULL )
    {
        e = new Edge( *v.AdjacentEdge );
    }

    delete AdjacentEdge;

    Cost = cost;
    Id = name;
    AdjacentEdge = e;
    Parent = p;
    return *this;
}

bool Vertex::operator==( const Vertex &v ) const
{
    if ( Id != v.Id )
    {
        return false;
    }

    if ( Cost != v.Cost )
    {
        return false;
    }

    return true;
}

bool Vertex::operator!=( const Vertex &v )const
{
    return !operator==( v );
}
